R/dev-mode.R

```diff
@@ -0,0 +1,243 @@
+#     message(
+#       pre_txt, "\n",
+#       paste0(
+#         "  ", opts_txt, desc_txt,
+#         collapse = "\n"
+#       )
+#     )
+#   }
+
+#   # return the return value when setting the `options()`
+#   options(opts_to_set)
+# }
+
+# ------------------------------------------
+
+
+#' Shiny Developer Mode
+#'
+#' @keywords internal
+#' @describeIn shiny_dev_mode A small function to set two options to enable/disable Shiny Developer Mode and whether or not to display Developer messages
+#' @param dev_mode Logical value which should be set to TRUE to enable Shiny Developer Mode
+#' @param verbose Logical value which should be set to display Shiny Developer messages
+#' @export
+#' @examples
+#' # Enable Shiny Developer mode
+#' shiny_dev_mode()
+shiny_dev_mode <- function(
+  dev_mode = getOption("shiny.devmode", TRUE),
+  verbose = getOption("shiny.devmode.verbose", TRUE)
+) {
+  options(
+    shiny.devmode = dev_mode,
+    shiny.devmode.verbose = verbose
+  )
+}
+
+
+#' @describeIn shiny_dev_mode Determines if Shiny is in Developer Mode
+#' @export
+#' @examples
+#' in_shiny_dev_mode() # TRUE/FALSE?
+in_shiny_dev_mode <- function() {
+  isTRUE(getOption("shiny.devmode", FALSE)) &&
+    # !testthat::is_testing()
+    !identical(Sys.getenv("TESTTHAT"), "true")
+}
+
+# ' @describeIn shiny_dev_mode Determines if Shiny Developer Mode messages should be displayed. Each message is displayed only once every 8 hours
+# ' @examples
+# ' is_shiny_dev_mode_verbose() # TRUE/FALSE?
+### Do not document
+is_shiny_dev_mode_verbose <- function() {
+  in_shiny_dev_mode() && isTRUE(getOption("shiny.devmode.verbose", TRUE))
+}
+
+
+#' @describeIn shiny_dev_mode Temporarily set Shiny Developer Mode and Developer message verbosity
+#' @param code Code to execute with the temporary `options()` set
+#' @export
+#' @examples
+#' # Execute code in a temporary shiny dev mode
+#' with_shiny_dev_mode(TRUE, in_shiny_dev_mode()) # TRUE
+with_shiny_dev_mode <- function(
+  dev_mode,
+  code,
+  verbose = getOption("shiny.devmode.verbose", TRUE)
+) {
+  with_options(
+    list(
+      shiny.devmode = dev_mode,
+      shiny.devmode.verbose = verbose
+    ),
+    code
+  )
+}
+
+#' @describeIn  shiny_dev_mode Helper method to retrieve an `options()` value. If the `option` is not set, `on_value` or `off_value` will be returned depending on the result of `in_shiny_dev_mode()`.  When Shiny Developer Mode is on, `on_msg` will be displayed using `shiny_dev_mode_msg()`
+#' @param option Option to look for in `options()`
+#' @param on_msg Message to display once every 8 hours if `option` is not set and `in_shiny_dev_mode()` returns `TRUE`.
+#' @param on_value Default value to use when `option` is not set and `in_shiny_dev_mode()` returns `TRUE`.
+#' @param off_value Default value to use when `option` is not set and `in_shiny_dev_mode()` returns `FALSE`.
+#' @export
+#' @examples
+#'
+#' # Used within `shiny::runApp(launch.browser)`
```

Comment:
Let's not set `shiny.launch.browser` because it doesn't seem like a dev mode related setting.